### PR TITLE
fix: check trunkENI attached status if it already exists before adding to cache

### PR DESCRIPTION
**Pull request by sushrk**
*Issue #, if available:*
#209 
*Description of changes:*
1. Check attached status if the trunkENI exists before adding to cache to avoid the leaked ENI breaking SGPP communication. 
2. Refactored InitTrunk test case to use table-driven test cases
3. Bump gomega and update developer guide

Testing 
```
SGPP tests
Ran 19 of 23 Specs in 1732.780 seconds
FAIL! -- 18 Passed | 1 Failed | 0 Pending | 4 Skipped

Webhook tests
Ran 11 of 11 Specs in 146.382 seconds
SUCCESS! -- 11 Passed | 0 Failed | 0 Pending | 0 Skipped

```
Known failure fixed in [commit ba8963a](https://github.com/aws/amazon-vpc-resource-controller-k8s/pull/286/commits/ba8963a43af613e0f6f280890bfc6456a44703b3), re-ran with focus- 

```
Branch ENI Pods Toggle Node between Managed/Un-Managed when node is toggled from managed to un-managed and back to managed pod should not run when un-managed and run when managed

Ran 1 of 23 Specs in 546.345 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 22 Skipped
```

Ran scale up/down tests with upto 50 nodes and 100 pods and verified branch & trunkeni is allocated and deleted as expected.
```
{"level":"info","timestamp":"2023-08-10T20:59:41.269Z","logger":"branch eni provider","msg":"created and annotated branch interface/s successfully","pod namespace":"default","pod name":"hello-world-7db65f56dd-xg6vt","node name":"XX","branches":[{"eniId":"eni-XX","ifAddress":"XX","privateIp":"XX","vlanId":X,"subnetCidr":"XX"}]}
{"level":"info","timestamp":"2023-08-10T21:07:05.003Z","logger":"branch eni provider","msg":"pushing ENIs to delete queue","node name":"XX","ENIs":[{"eniId":"eni-XX","ifAddress":"XX","privateIp":"XX","vlanId":X,"subnetCidr":"XX"}]}

{"level":"info","timestamp":"2023-08-10T20:59:33.123Z","logger":"branch eni provider","msg":"trunk added to cache successfully","node":"XX"}
{"level":"info","timestamp":"2023-08-10T21:11:16.003Z","logger":"node manager","msg":"node removed from data store","node name":"XX","request":"delete"}
{"level":"info","timestamp":"2023-08-10T20:35:21.971Z","logger":"branch eni provider","msg":"trunk removed from cache successfully","node":"XX"}

{"level":"info","timestamp":"2023-08-10T21:17:47.428Z","logger":"branch eni provider","msg":"trunk removed from cache successfully","node":"XX"}
{"level":"info","timestamp":"2023-08-10T21:17:47.428Z","logger":"branch eni provider","msg":"de-initialized resource provider successfully","node name":"XX"}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/aws/ec2/api/helper.go [hunk 1]
```diff
@@ -607,7 +607,7 @@ func (h *ec2APIHelper) DetachAndDeleteNetworkInterface(attachmentID *string, nwI
 	if err != nil {
 		return err
 	}
-	err = h.WaitForNetworkInterfaceStatusChange(nwInterfaceID, ec2.NetworkInterfaceStatusAvailable)
+	err = h.WaitForNetworkInterfaceStatusChange(nwInterfaceID, ec2.AttachmentStatusDetached)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (sushrk) [reply pkg/aws/ec2/api/helper.go:611]: `WaitForNetworkInterfaceStatusChange` actually checks the NI attachment status, and the valid values are `AttachmentStatusAttaching("attaching")`, `AttachmentStatusAttached("attached")`, `AttachmentStatusDetaching("detaching")`, `AttachmentStatusDetached("detached")`. 

Maybe good to also update the function name to something like `WaitForNetworkInterfaceAtchStatusChange`?

[the remark this answers is not in the document]
Comment (haouc) [reply pkg/aws/ec2/api/helper.go:611]: The `Available` is a Go-SDK defined enum
```
const (
	// NetworkInterfaceStatusAvailable is a NetworkInterfaceStatus enum value
	NetworkInterfaceStatusAvailable = "available"
```
https://raw.githubusercontent.com/aws/aws-sdk-go/main/service/ec2/api.go
I am more concern on which status we should `Wait`for. `Available` and `Detached` are different.

[the remark this answers is not in the document]
Comment (sushrk) [reply pkg/aws/ec2/api/helper.go:611]: This is the enum for the network interface status; `Available` is not a valid value for the network interface _attachment_ status. Do you mean we should be checking the network interface status instead?

[the remark this answers is not in the document]
Comment (haouc) [reply pkg/aws/ec2/api/helper.go:611]: Got it. I would prefer to align all our deleting standard to be same, either `Available` or `Detached`. Maybe `Available` would be safer?

File: pkg/provider/branch/trunk/trunk.go [hunk 2]
```diff
@@ -174,14 +174,19 @@ func (t *trunkENI) InitTrunk(instance ec2.EC2Instance, podList []v1.Pod) error {
...
-		// It's possible to get an empty network interface response if the instnace
-		// is being deleted.
+		// It's possible to get an empty network interface response if the instance is being deleted.
 		if nwInterface == nil || nwInterface.InterfaceType == nil {
 			return fmt.Errorf("received an empty network interface response "+
 				"from EC2 %+v", nwInterface)
 		}
 		if *nwInterface.InterfaceType == "trunk" {
-			t.trunkENIId = *nwInterface.NetworkInterfaceId
+			fmt.Println("checking the attachment status")
```
Comment (haouc) [inline pkg/provider/branch/trunk/trunk.go]: This seems for debug?

**Review by haouc — APPROVED** [review]
lgtm

File: pkg/aws/ec2/api/helper.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sushrk) [reply pkg/aws/ec2/api/helper.go:611]: To close the loop on this- synced offline, we will revisit this to have similar checks across all clean up logic.

**Pull request merged**